mesonbuild/interpreter.py

```diff
@@ -3169,9 +3169,24 @@ def func_project(self, node, args, kwargs):
         if not self.is_subproject():
             self.build.project_name = proj_name
         self.active_projectname = proj_name
-        self.project_version = kwargs.get('version', 'undefined')
-        if not isinstance(self.project_version, str):
-            raise InvalidCode('The version keyword argument must be a string.')
+        version = kwargs.get('version', 'undefined')
+        if isinstance(version, list):
+            if len(version) != 1:
+                raise InvalidCode('Version argument is an array with more than one entry.')
+            version = version[0]
+        if isinstance(version, mesonlib.File):
+            self.add_build_def_file(version)
+            ifname = version.absolute_path(self.environment.source_dir,
+                                           self.environment.build_dir)
+            try:
+                with open(ifname, 'r') as versionfile:
+                    self.project_version = versionfile.readline().strip()
```

Comment:
Should we be allowing version files that have additional non-version lines in it? Yes, it's documented as such, but there does not seem to be a good reason to allow it, and IMHO it is confusing.

Reply:
Either one is fine by me.

Reply:
For mesa our version file is: `21.0.0-rc4\n`. I'd prefer to error if there's more than one line, we can always change it later if there's a compelling reason, and it's a pretty trivial change for a project if they have more than one line.